--- ./KernelSU/kernel/apk_sign.c.orig	2024-12-18 08:53:14.063525500 -0500
+++ ./KernelSU/kernel/apk_sign.c	2024-12-18 09:10:04.227156884 -0500
@@ -314,8 +314,14 @@
 
 #endif
 
-bool is_manager_apk(char *path)
+bool ksu_is_manager_apk(char *path)
 {
+#ifdef CONFIG_KSU_SUSFS
+	return (check_v2_signature(path, EXPECTED_SIZE, EXPECTED_HASH) 
+	|| check_v2_signature(path, EXPECTED_NEXT_SIZE, EXPECTED_NEXT_HASH)
+	|| check_v2_signature(path, EXPECTED_MAGIC_SIZE, EXPECTED_MAGIC_HASH));
+#else
 	return (check_v2_signature(path, EXPECTED_SIZE, EXPECTED_HASH) 
 	|| check_v2_signature(path, EXPECTED_NEXT_SIZE, EXPECTED_NEXT_HASH));
+#endif
 }
\ No newline at end of file
