--- ./KernelSU-Next/kernel/apk_sign.c	2024-12-18 08:53:14.063525500 -0500
+++ ./KernelSU-Next/kernel/apk_sign.c	2024-12-18 09:10:04.227156884 -0500
@@ -314,8 +314,18 @@
 
 #endif
 
-bool is_manager_apk(char *path)
+bool ksu_is_manager_apk(char *path)
 {
+#ifdef CONFIG_KSU_SUSFS
+	return (check_v2_signature(path, EXPECTED_SIZE, EXPECTED_HASH) 
+	|| check_v2_signature(path, EXPECTED_NEXT_SIZE, EXPECTED_NEXT_HASH)
+	|| check_v2_signature(path, EXPECTED_MAGIC_SIZE, EXPECTED_MAGIC_HASH)
+ || check_v2_signature(path, 0x363, "4359c171f32543394cbc23ef908c4bb94cad7c8087002ba164c8230948c21549") // backslashxx
+ || check_v2_signature(path, 384, "7e0c6d7278a3bb8e364e0fcba95afaf3666cf5ff3c245a3b63c8833bd0445cc4") // 5ec1cff
+ || check_v2_signature(path, 0x396, "f415f4ed9435427e1fdf7f1fccd4dbc07b3d6b8751e4dbcec6f19671f427870b") // rsuntk
+ );
+#else
 	return (check_v2_signature(path, EXPECTED_SIZE, EXPECTED_HASH) 
 	|| check_v2_signature(path, EXPECTED_NEXT_SIZE, EXPECTED_NEXT_HASH));
+#endif
 }
\ No newline at end of file
