--- a/include/linux/sched.h	2024-12-15 03:41:21.349675000 -0500
+++ b/include/linux/sched.h	2024-12-15 03:41:52.759768540 -0500
@@ -1521,6 +1521,10 @@
 	ANDROID_KABI_RESERVE(7);
 	ANDROID_KABI_RESERVE(8);
 
+#ifdef CONFIG_KSU_SUSFS
+	u64 android_kabi_reserved8;
+#endif
+
 	/*
 	 * New fields for task_struct should be added above here, so that
 	 * they are included in the randomized portion of task_struct.
