From e0169787aafa148a55e25bdd3aae004980084c00 Mon Sep 17 00:00:00 2001
From: fatalcoder524 <11532648+fatalcoder524@users.noreply.github.com>
Date: Tue, 7 Jul 2026 08:32:46 +0000
Subject: [PATCH] Enable volume input to enable bbg efisp patch

---
 anykernel.sh | 65 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 65 insertions(+)

diff --git a/anykernel.sh b/anykernel.sh
index 6a97197..5360e0b 100755
--- a/anykernel.sh
+++ b/anykernel.sh
@@ -51,6 +51,71 @@ $ksu_supported || abort "  -> Non-GKI device, abort."
 # boot install
 split_boot
 
+# Menu to conditionally patch BBG for abl/efisp whitelisting
+##################################################
+SELECTED_STR="  [->] Selected:"
+
+show_menu() {
+  local selected=1
+  local total=$#
+  if [ $total -eq 2 ]; then
+    while true; do
+      handle_input
+      KEY_RESULT=$(cat /tmp/ak3_key_result 2>/dev/null)
+      case "$KEY_RESULT" in
+      "KEY_VOLUMEUP") return 1 ;;
+      "KEY_VOLUMEDOWN"|"TIMEOUT") return 2 ;;
+      esac
+    done
+  else
+    while true; do
+      eval "local current=\"\$$selected\""
+      ui_print "➔ $current"
+      ui_print " "
+      handle_input
+      KEY_RESULT=$(cat /tmp/ak3_key_result 2>/dev/null)
+      case "$KEY_RESULT" in
+      "KEY_VOLUMEUP") 
+		selected=$((selected % total + 1))
+		;;
+      "KEY_VOLUMEDOWN"|"TIMEOUT")
+        ui_print "$SELECTED_STR $current"
+        return $selected
+        ;;
+      esac
+    done
+  fi
+}
+
+MENU_OPTIONS="Yes No"
+
+PATCH_CMDLINE=0
+ui_print " "
+ui_print "=========================================="
+ui_print " Are you using efisp exploit and require "
+ui_print " whitelisting abl and efisp partitions?   "
+ui_print "------------------------------------------"
+ui_print " 1. Yes. Whitelist abl and efisp!        "
+ui_print " 2. No. Keep as is!                      "
+ui_print "                                         "
+ui_print "        [VOL+] - Yes | [VOL-] - No       "
+ui_print "=========================================="
+ui_print " "
+show_menu $MENU_OPTIONS
+case $? in
+    1) PATCH_CMDLINE=1 ;;
+    2) PATCH_CMDLINE=0 ;;
+esac
+
+if [ $PATCH_CMDLINE == 1 ]; then
+    ui_print "Patching cmdline to allow abl and efisp flashing!!!"
+    patch_cmdline "oplusboot.secure_user_mode" "oplusboot.secure_user_mode=0"
+else
+    ui_print "Flashing the kernel as is with full BBG security!!!"
+    patch_cmdline "oplusboot.secure_user_mode" ""
+fi
+##################################################
+
 if [ -f "$SPLITIMG/ramdisk.cpio" ]; then
     unpack_ramdisk
     write_boot
-- 
2.47.3

