diff --git a/kernel/kernel_compat.c b/kernel/kernel_compat.c
index 873da944..267d150c 100644
--- ./kernel/kernel_compat.c
+++ ./kernel/kernel_compat.c
@@ -182,19 +182,10 @@ long ksu_strncpy_from_user_nofault(char *dst, const void __user *unsafe_addr,
 
 	return ret;
 }
 #endif
 
-static inline int ksu_access_ok(const void *addr, unsigned long size)
-{
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,0,0)
-	return access_ok(addr, size);
-#else
-	return access_ok(VERIFY_READ, addr, size);
-#endif
-}
-
 long ksu_strncpy_from_user_retry(char *dst, const void __user *unsafe_addr,
 				   long count)
 {
 	long ret = ksu_strncpy_from_user_nofault(dst, unsafe_addr, count);
 	if (likely(ret >= 0))
