diff --git a/kernel/ksud.c b/kernel/ksud.c
index 11f865df..8f3d565f 100644
--- a/kernel/ksud.c
+++ b/kernel/ksud.c
@@ -25,11 +25,16 @@
 #include "ksud.h"
 #include "selinux/selinux.h"
 #include "throne_tracker.h"
-#include "hook/syscall_hook.h"
 
 bool ksu_module_mounted __read_mostly = false;
 bool ksu_boot_completed __read_mostly = false;
 
+#ifdef CONFIG_KSU_SUSFS
+bool ksu_init_rc_hook __read_mostly = true;
+bool ksu_execveat_hook __read_mostly = true;
+bool ksu_input_hook __read_mostly = true;
+#endif // #ifdef CONFIG_KSU_SUSFS
+
 static const char KERNEL_SU_RC[] =
 	"\n"
 
@@ -103,28 +108,13 @@ void on_module_mounted(void)
 	ksu_module_mounted = true;
 }
 
-extern void ksu_avc_spoof_late_init();
 void on_boot_completed(void)
 {
     ksu_boot_completed = true;
     pr_info("on_boot_completed!\n");
     track_throne(true);
-    ksu_avc_spoof_late_init();
 }
 
-#define MAX_ARG_STRINGS 0x7FFFFFFF
-struct user_arg_ptr {
-#ifdef CONFIG_COMPAT
-	bool is_compat;
-#endif
-	union {
-		const char __user *const __user *native;
-#ifdef CONFIG_COMPAT
-		const compat_uptr_t __user *compat;
-#endif
-	} ptr;
-};
-
 static const char __user *get_user_arg_ptr(struct user_arg_ptr argv, int nr)
 {
 	const char __user *native;
@@ -194,7 +184,7 @@ static bool check_argv(struct user_arg_ptr argv, int index,
 	if (!p || IS_ERR(p))
 		goto fail;
 
-	if (strncpy_from_user_nofault(buf, p, buf_len) <= 0)
+	if (strncpy_from_user(buf, p, buf_len) <= 0)
 		goto fail;
 
 	buf[buf_len - 1] = '\0';
@@ -205,8 +195,17 @@ fail:
 	return false;
 }
 
-void ksu_handle_execveat_ksud(const char *path, struct user_arg_ptr *argv)
+#ifdef CONFIG_KSU_SUSFS
+extern int ksu_handle_execveat_init(struct filename *filename);
+#endif // #ifdef CONFIG_KSU_SUSFS
+
+// IMPORTANT NOTE: the call from execve_handler_pre WON'T provided correct value for envp and flags in GKI version
+int ksu_handle_execveat_ksud(int *fd, struct filename **filename_ptr,
+                             struct user_arg_ptr *argv,
+                             struct user_arg_ptr *envp, int *flags)
 {
+    struct filename *filename;
+
 	static const char app_process[] = "/system/bin/app_process";
 	static bool first_zygote = true;
 
@@ -214,9 +213,18 @@ void ksu_handle_execveat_ksud(const char *path, struct user_arg_ptr *argv)
 	static const char system_bin_init[] = "/system/bin/init";
 	static bool init_second_stage_executed = false;
 
+    if (!filename_ptr)
+        return 0;
+
+    filename = *filename_ptr;
+    if (IS_ERR(filename)) {
+        return 0;
+    }
+
 	// https://cs.android.com/android/platform/superproject/+/android-16.0.0_r2:system/core/init/main.cpp;l=77
-	if (unlikely(!memcmp(path, system_bin_init, sizeof(system_bin_init) - 1) &&
-			argv)) {
+    if (unlikely(!memcmp(filename->name, system_bin_init,
+                         sizeof(system_bin_init) - 1) &&
+                 argv)) {
 		char buf[16];
 		if (!init_second_stage_executed &&
 			check_argv(*argv, 1, "second_stage", buf, sizeof(buf))) {
@@ -228,17 +236,26 @@ void ksu_handle_execveat_ksud(const char *path, struct user_arg_ptr *argv)
 		}
 	}
 
-	if (unlikely(first_zygote &&
-			!memcmp(path, app_process, sizeof(app_process) - 1) && argv)) {
+    if (unlikely(first_zygote &&
+            !memcmp(filename->name, app_process, sizeof(app_process) - 1) &&
+            argv))
+    {
 		char buf[16];
 		if (check_argv(*argv, 1, "-Xzygote", buf, sizeof(buf))) {
 			pr_info("exec zygote, /data prepared, second_stage: %d\n",
 				init_second_stage_executed);
 			on_post_fs_data();
 			first_zygote = false;
-			ksu_stop_ksud_execve_hook();
+			stop_execveat_hook();
 		}
 	}
+
+#ifdef CONFIG_KSU_SUSFS
+    // - We need to run ksu_handle_execveat_init() at the very end in case the above checks are skipped
+    (void)ksu_handle_execveat_init(filename);
+#endif // #ifdef CONFIG_KSU_SUSFS
+
+    return 0;
 }
 
 static ssize_t (*orig_read)(struct file *, char __user *, size_t, loff_t *);
@@ -385,7 +402,7 @@ static void ksu_install_rc_hook(struct file *file)
 	file->f_op = &fops_proxy;
 }
 
-static void ksu_handle_sys_read(unsigned int fd, char __user **buf_ptr,
+void ksu_handle_sys_read(unsigned int fd, char __user **buf_ptr,
 				size_t *count_ptr)
 {
 	struct file *file = fget(fd);
@@ -406,6 +423,11 @@ static bool is_volumedown_enough(unsigned int count)
 int ksu_handle_input_handle_event(unsigned int *type, unsigned int *code,
 					int *value)
 {
+#ifdef CONFIG_KSU_SUSFS
+    if (!ksu_input_hook)
+        return 0;
+#endif // #ifdef CONFIG_KSU_SUSFS
+
 	if (*type == EV_KEY && *code == KEY_VOLUMEDOWN) {
 		int val = *value;
 		pr_info("KEY_VOLUMEDOWN val: %d\n", val);
@@ -429,10 +451,6 @@ bool ksu_is_safe_mode()
 		return true;
 	}
 
-	if (ksu_late_loaded) {
-		return false;
-	}
-
 	// stop hook first!
 	stop_input_hook();
 
@@ -447,138 +465,53 @@ bool ksu_is_safe_mode()
 	return false;
 }
 
-void ksu_execve_hook_ksud(const struct pt_regs *regs)
-{
-	const char __user **filename_user = (const char **)&PT_REGS_PARM1(regs);
-	const char __user *const __user *__argv =
-		(const char __user *const __user *)PT_REGS_PARM2(regs);
-	struct user_arg_ptr argv = { .ptr.native = __argv };
-	char path[32];
-	long ret;
-	unsigned long addr;
-	const char __user *fn;
-
-	if (!filename_user)
-		return;
-
-	addr = untagged_addr((unsigned long)*filename_user);
-	fn = (const char __user *)addr;
+#ifdef CONFIG_KSU_SUSFS
+void ksu_handle_vfs_fstat(int fd, loff_t *kstat_size_ptr) {
+    loff_t new_size = *kstat_size_ptr + ksu_rc_len;
+    struct file *file = fget(fd);
 
-	memset(path, 0, sizeof(path));
-	ret = strncpy_from_user(path, fn, 32);
-	if (ret < 0) {
-		pr_err("Access filename failed for execve_handler_pre\n");
+	if (!file)
 		return;
-	}
-
-	ksu_handle_execveat_ksud(path, &argv);
-}
-
-static long (*orig_sys_read)(const struct pt_regs *regs);
-static long ksu_sys_read(const struct pt_regs *regs)
-{
-	unsigned int fd = PT_REGS_PARM1(regs);
-	char __user **buf_ptr = (char __user **)&PT_REGS_PARM2(regs);
-	size_t *count_ptr = (size_t *)&PT_REGS_PARM3(regs);
 
-	ksu_handle_sys_read(fd, buf_ptr, count_ptr);
-	return orig_sys_read(regs);
-}
-
-static long (*orig_sys_fstat)(const struct pt_regs *regs);
-static long ksu_sys_fstat(const struct pt_regs *regs)
-{
-	unsigned int fd = PT_REGS_PARM1(regs);
-	void __user *statbuf = (void __user *)PT_REGS_PARM2(regs);
-	bool is_rc = false;
-	long ret;
-
-	struct file *file = fget(fd);
-	if (file) {
-		if (is_init_rc(file)) {
-			pr_info("stat init.rc");
-			is_rc = true;
-		}
-		fput(file);
+	if (is_init_rc(file)) {
+        pr_info("stat init.rc");
+        pr_info("adding ksu_rc_len: %lld -> %lld", *kstat_size_ptr, new_size);
+        *kstat_size_ptr = new_size;
 	}
 
-	ret = orig_sys_fstat(regs);
-
-	if (is_rc) {
-		void __user *st_size_ptr = statbuf + offsetof(struct stat, st_size);
-		long size, new_size;
-		if (!copy_from_user_nofault(&size, st_size_ptr, sizeof(long))) {
-			new_size = size + ksu_rc_len;
-			pr_info("adding ksu_rc_len: %ld -> %ld", size, new_size);
-			if (!copy_to_user_nofault(st_size_ptr, &new_size, sizeof(long))) {
-				pr_info("added ksu_rc_len");
-			} else {
-				pr_err("add ksu_rc_len failed: statbuf 0x%lx",
-					(unsigned long)st_size_ptr);
-			}
-		} else {
-			pr_err("read statbuf 0x%lx failed", (unsigned long)st_size_ptr);
-		}
-	}
-
-	return ret;
-}
-
-static int input_handle_event_handler_pre(struct kprobe *p,
-						struct pt_regs *regs)
-{
-	unsigned int *type = (unsigned int *)&PT_REGS_PARM2(regs);
-	unsigned int *code = (unsigned int *)&PT_REGS_PARM3(regs);
-	int *value = (int *)&PT_REGS_CCALL_PARM4(regs);
-	return ksu_handle_input_handle_event(type, code, value);
+	fput(file);
 }
+#endif // #ifdef CONFIG_KSU_SUSFS
 
-static struct kprobe input_event_kp = {
-	.symbol_name = "input_event",
-	.pre_handler = input_handle_event_handler_pre,
-};
-
-static void do_stop_input_hook(struct work_struct *work)
+static void stop_init_rc_hook()
 {
-	unregister_kprobe(&input_event_kp);
+    ksu_init_rc_hook = false;
+    pr_info("stop_init_rc_hook()\n");
 }
 
-static void stop_init_rc_hook()
+static void stop_execveat_hook()
 {
-	ksu_syscall_table_unhook(__NR_read);
-	ksu_syscall_table_unhook(__NR_fstat);
-	pr_info("unregister init_rc syscall hook\n");
+	ksu_execveat_hook = false;
+    pr_info("stop_execveat_hook()\n");
 }
 
 static void stop_input_hook()
 {
 	static bool input_hook_stopped = false;
-	if (input_hook_stopped) {
-		return;
-	}
+    if (input_hook_stopped)
+         return;
+    ksu_input_hook = false;
 	input_hook_stopped = true;
-	bool ret = schedule_work(&stop_input_hook_work);
-	pr_info("unregister input kprobe: %d!\n", ret);
+    pr_info("stop_input_hook()\n");
 }
 
 // ksud: module support
 void ksu_ksud_init()
 {
-	int ret;
-
-	ksu_syscall_table_hook(__NR_read, ksu_sys_read, &orig_sys_read);
-	ksu_syscall_table_hook(__NR_fstat, ksu_sys_fstat, &orig_sys_fstat);
 
-	ret = register_kprobe(&input_event_kp);
-	pr_info("ksud: input_event_kp: %d\n", ret);
-
-	INIT_WORK(&stop_input_hook_work, do_stop_input_hook);
 }
 
 void ksu_ksud_exit()
 {
-	// TODO:
-	// this should be done before unregister vfs_read_kp
-	// stop_init_rc_hook();
-	unregister_kprobe(&input_event_kp);
+
 }
