diff --git a/kernel/hook/setuid_hook.c b/kernel/hook/setuid_hook.c
index ff232021..9e2bc15d 100644
--- a/kernel/hook/setuid_hook.c
+++ b/kernel/hook/setuid_hook.c
@@ -20,40 +20,66 @@
 #include "manager/manager_identity.h"
 #include "infra/seccomp_cache.h"
 #include "supercall/supercall.h"
-#include "hook/tp_marker.h"
 #include "feature/kernel_umount.h"
 
-int ksu_handle_setresuid(uid_t old_uid, uid_t new_uid)
+extern u32 susfs_zygote_sid;
+extern void disable_seccomp(void);
+extern struct work_struct susfs_extra_works;
+
+static inline void ksu_handle_extra_susfs_work(void)
 {
-    // we rely on the fact that zygote always call setresuid(3) with same uids
+    if (work_pending(&susfs_extra_works))
+        return;
+
+    schedule_work(&susfs_extra_works);
+}
 
-    pr_info("handle_setresuid from %d to %d\n", old_uid, new_uid);
+int ksu_handle_setresuid(uid_t ruid, uid_t euid, uid_t suid)
+{
+    // We only interest in process spwaned by zygote
+    if (!susfs_is_sid_equal(current_cred(), susfs_zygote_sid))
+        return 0;
 
-    if (unlikely(is_uid_manager(new_uid))) {
-        spin_lock_irq(&current->sighand->siglock);
-        ksu_seccomp_allow_cache(current->seccomp.filter, __NR_reboot);
-        ksu_set_task_tracepoint_flag(current);
-        spin_unlock_irq(&current->sighand->siglock);
+    // Check if spawned process is isolated service first, and force to do umount if so
+    if (is_isolated_process(ruid))
+        goto do_umount;
 
-        pr_info("install fd for manager: %d\n", new_uid);
+    // - Since ksu maanger app uid is excluded in allow_list_arr, so ksu_uid_should_umount(manager_uid)
+    //   will always return true, that's why we need to explicitly check if new_uid belongs to
+    //   ksu manager.
+    // - Disable seccomp restriction for KSU manager since running with "su" will disable seccomp anyway
+    if (likely(ksu_is_manager_appid_valid()) && unlikely(is_uid_manager(ruid))) {
+        disable_seccomp();
+        pr_info("install fd for manager: %d\n", ruid);
         ksu_install_fd();
         return 0;
     }
 
-    if (ksu_is_allow_uid_for_current(new_uid)) {
-        if (current->seccomp.mode == SECCOMP_MODE_FILTER &&
-            current->seccomp.filter) {
-            spin_lock_irq(&current->sighand->siglock);
-            ksu_seccomp_allow_cache(current->seccomp.filter, __NR_reboot);
-            spin_unlock_irq(&current->sighand->siglock);
-        }
-        ksu_set_task_tracepoint_flag(current);
-    } else {
-        ksu_clear_task_tracepoint_flag_if_needed(current);
+    // we should not umount for webview zygote
+    if (unlikely(ruid == WEBVIEW_ZYGOTE_UID))
+        return 0;
+
+    // Check if spawned process is normal user app and needs to be umounted
+    if (likely(is_appuid(ruid) && ksu_uid_should_umount(ruid)))
+        goto do_umount;
+
+    // - Disable seccomp restriction for root allowed apps since running with "su" will disable seccomp anyway
+    if (ksu_is_allow_uid_for_current(ruid))
+        disable_seccomp();
+
+    return 0;
+
+do_umount:
+    {
+        // Handle kernel umount
+        ksu_handle_umount(current_uid().val, ruid);
+
+        // Handle extra susfs work
+        ksu_handle_extra_susfs_work();
     }
 
-    // Handle kernel umount
-    ksu_handle_umount(old_uid, new_uid);
+    // Mark current proc as umounted
+    susfs_set_current_proc_umounted();
 
     return 0;
 }
