diff --git a/kernel/Kbuild b/kernel/Kbuild
index 30b676b6..d5bfcd66 100644
--- a/kernel/Kbuild
+++ b/kernel/Kbuild
@@ -13,13 +13,6 @@ kernelsu-objs += feature.o
 kernelsu-objs += ksud.o
 kernelsu-objs += seccomp_cache.o
 kernelsu-objs += file_wrapper.o
-ifeq ($(CONFIG_ARM64),y)
-kernelsu-objs += hook/arm64/patch_memory.o
-kernelsu-objs += hook/arm64/syscall_hook.o
-else ifeq ($(CONFIG_X86_64),y)
-kernelsu-objs += hook/x86_64/patch_memory.o
-kernelsu-objs += hook/x86_64/syscall_hook.o
-endif
 kernelsu-objs += extras.o
 
 kernelsu-objs += selinux/selinux.o
