diff --git a/kernel/hook/setuid_hook.c b/kernel/hook/setuid_hook.c
index d3aef3c1..238a9e4e 100644
--- a/kernel/hook/setuid_hook.c
+++ b/kernel/hook/setuid_hook.c
@@ -81,6 +81,11 @@ int ksu_handle_setresuid(uid_t ruid, uid_t euid, uid_t suid)
         return 0;
     }
 
+    // Check if spawned process is normal user app and needs to be umounted
+    if (likely(is_zygote_normal_app_uid(new_uid) && ksu_uid_should_umount(new_uid)))
+        goto do_umount;
+
+
     if (ksu_is_allow_uid_for_current(new_uid)) {
         if (current->seccomp.mode == SECCOMP_MODE_FILTER &&
             current->seccomp.filter) {
@@ -88,14 +93,20 @@ int ksu_handle_setresuid(uid_t ruid, uid_t euid, uid_t suid)
             ksu_seccomp_allow_cache(current->seccomp.filter, __NR_reboot);
             spin_unlock_irq(&current->sighand->siglock);
         }
-        ksu_set_task_tracepoint_flag(current);
-    } else {
-        ksu_clear_task_tracepoint_flag_if_needed(current);
     }
 
+    return 0;
+
+do_umount:
     // Handle kernel umount
     ksu_handle_umount(old_uid, new_uid);
 
+    // Handle extra susfs work
+    ksu_handle_extra_susfs_work();
+
+    // Mark current proc as umounted
+    susfs_set_current_proc_umounted();
+
     return 0;
 }
 
