--- a/kernel/feature/sucompat.c
+++ b/kernel/feature/sucompat.c
@@ -22,6 +22,7 @@
 #include "klog.h" // IWYU pragma: keep
 #include "runtime/ksud.h"
 #include "feature/sucompat.h"
+#include "feature/adb_root.h"
 #include "policy/app_profile.h"
 #include "hook/syscall_hook.h"
 
@@ -30,18 +31,28 @@
 #define SU_PATH "/system/bin/su"
 #define SH_PATH "/system/bin/sh"
 
-bool ksu_su_compat_enabled __read_mostly = true;
+static const char sh_path[] = SH_PATH;
+static const char su_path[] = SU_PATH;
+static const char ksud_path[] = KSUD_PATH;
+
+DEFINE_STATIC_KEY_TRUE(ksu_su_compat_enabled);
 
 static int su_compat_feature_get(u64 *value)
 {
-	*value = ksu_su_compat_enabled ? 1 : 0;
+    if (static_key_enabled(&ksu_su_compat_enabled)) {
+        *value = 1;
+    } else {
+        *value = 0; }
 	return 0;
 }
 
 static int su_compat_feature_set(u64 value)
 {
 	bool enable = value != 0;
-	ksu_su_compat_enabled = enable;
+    if (enable) {
+        static_branch_enable(&ksu_su_compat_enabled);
+    } else {
+        static_branch_disable(&ksu_su_compat_enabled); }
 	pr_info("su_compat: set to %d\n", enable);
 	return 0;
 }
@@ -76,103 +87,154 @@
 	return userspace_stack_buffer(ksud_path, sizeof(ksud_path));
 }
 
-int ksu_handle_faccessat(int *dfd, const char __user **filename_user,
-		int *mode, int *__unused_flags)
+static const char __user *get_user_arg_ptr(struct user_arg_ptr argv, int nr)
+{
+	const char __user *native;
+
+#ifdef CONFIG_COMPAT
+	if (unlikely(argv.is_compat)) {
+		compat_uptr_t compat;
+
+		if (get_user(compat, argv.ptr.compat + nr))
+			return ERR_PTR(-EFAULT);
+
+		return compat_ptr(compat);
+	}
+#endif
+
+	if (get_user(native, argv.ptr.native + nr))
+		return ERR_PTR(-EFAULT);
+
+	return native;
+}
+
+/*
+ * return 0 -> No further checks should be required afterwards
+ * return non-zero -> Further checks should be continued afterwards
+ */
+int ksu_handle_execveat_init(struct filename *filename, struct user_arg_ptr *argv_user, struct user_arg_ptr *envp_user) {
+    if (current->pid != 1 && is_init(get_current_cred())) {
+        int ret;
+        if (unlikely(strcmp(filename->name, KSUD_PATH) == 0)) {
+            pr_info("hook_manager: escape to root for init executing ksud: %d\n", current->pid);
+            ret = escape_to_root_for_init();
+            if (ret) {
+                pr_err("escape_to_root_for_init() failed: %d\n", ret);
+                return ret;
+            }
+            return 0;
+        } else if (likely(strstr(filename->name, "/app_process") == NULL &&
+                    strstr(filename->name, "/adbd") == NULL) &&
+                    !susfs_is_current_proc_umounted())
+        {
+            pr_info("susfs: mark no sucompat checks for pid: '%d', exec: '%s'\n", current->pid, filename->name);
+            susfs_set_current_proc_umounted();
+            return 0;
+        }
+
+#ifdef CONFIG_COMPAT
+        if (unlikely(envp_user->is_compat))
+            ret = ksu_adb_root_handle_execve(filename->name, (void ***)&envp_user->ptr.compat);
+        else
+            ret = ksu_adb_root_handle_execve(filename->name, (void ***)&envp_user->ptr.native);    
+#else
+        ret = ksu_adb_root_handle_execve(filename->name, (void ***)&envp_user->ptr.native);
+#endif
+
+        if (ret) {
+            pr_err("adb root failed: %d\n", ret);
+            return ret;
+        }
+        return ret;
+    }
+
+    return -EINVAL;
+}
+
+// the call from execve_handler_pre won't provided correct value for __never_use_argument, use them after fix execve_handler_pre, keeping them for consistence for manually patched code
+int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
+                 void *argv_user, void *envp_user,
+                 int *__never_use_flags)
 {
-	const char su[] = SU_PATH;
+    struct filename *filename;
+    int ret;
 
-	if (!ksu_is_allow_uid_for_current(current_uid().val)) {
+    if (unlikely(!filename_ptr))
 		return 0;
-	}
 
-	char path[sizeof(su) + 1];
-	memset(path, 0, sizeof(path));
-	strncpy_from_user_nofault(path, *filename_user, sizeof(path));
-
-	if (unlikely(!memcmp(path, su, sizeof(su)))) {
-		write_sulog('a');
-		pr_info("faccessat su->sh!\n");
-		*filename_user = sh_user_path();
-	}
+    filename = *filename_ptr;
+    if (IS_ERR(filename))
+        return 0;
+
+    if (!ksu_handle_execveat_init(filename, (struct user_arg_ptr*)argv_user, (struct user_arg_ptr*)envp_user))
+		return 0;
+
+    if (likely(memcmp(filename->name, su_path, sizeof(su_path))))
+        return 0;
+
+    pr_info("ksu_handle_execveat_sucompat: su found\n");
+
+    memcpy((void *)filename->name, ksud_path, sizeof(ksud_path));
+
+    ret = escape_with_root_profile();
+    if (ret)
+        pr_err("escape_with_root_profile() failed: %d\n", ret);
 
 	return 0;
 }
 
-int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags)
+int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv, void *envp, int *flags)
 {
-	// const char sh[] = SH_PATH;
-	const char su[] = SU_PATH;
+    if (ksu_handle_execveat_ksud(fd, filename_ptr, argv, envp, flags))
+        return 0;
 
-	if (!ksu_is_allow_uid_for_current(current_uid().val)) {
-		return 0;
-	}
+    return ksu_handle_execveat_sucompat(fd, filename_ptr, argv, envp, flags);
+}
 
-	if (unlikely(!filename_user)) {
-		return 0;
-	}
+int ksu_handle_faccessat(int *dfd, const char __user **filename_user, int *mode, int *__unused_flags)
+{
+    char path[sizeof(su_path) + 1] = {0};
 
-	char path[sizeof(su) + 1];
-	memset(path, 0, sizeof(path));
-	strncpy_from_user_nofault(path, *filename_user, sizeof(path));
-
-	if (unlikely(!memcmp(path, su, sizeof(su)))) {
-		write_sulog('s');
-		pr_info("newfstatat su->sh!\n");
-		*filename_user = sh_user_path();
+    strncpy_from_user(path, *filename_user, sizeof(path));
+
+    if (unlikely(!memcmp(path, su_path, sizeof(su_path)))) {
+        pr_info("ksu_handle_faccessat: su->sh!\n");
+        *filename_user = sh_user_path();
 	}
 
-	return 0;
+    return 0;
 }
 
-long ksu_handle_execve_sucompat(const char __user **filename_user, int orig_nr, const struct pt_regs *regs)
-{
-	const char su[] = SU_PATH;
-	const char __user *fn;
-	char path[sizeof(su) + 1];
-	long ret;
-	unsigned long addr;
-
-	if (unlikely(!filename_user))
-		goto do_orig_execve;
-
-	if (!ksu_is_allow_uid_for_current(current_uid().val))
-		goto do_orig_execve;
-
-	addr = untagged_addr((unsigned long)*filename_user);
-	fn = (const char __user *)addr;
-	memset(path, 0, sizeof(path));
-	ret = strncpy_from_user(path, fn, sizeof(path));
-
-	if (ret < 0) {
-		pr_warn("Access filename when execve failed: %ld", ret);
-		goto do_orig_execve;
-	}
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)
+int ksu_handle_stat(int *dfd, struct filename **filename, int *flags) {
+    if (unlikely(IS_ERR(*filename) || (*filename)->name == NULL))
+        return 0;
 
-	if (likely(memcmp(path, su, sizeof(su))))
-		goto do_orig_execve;
+    if (likely(memcmp((*filename)->name, su_path, sizeof(su_path))))
+        return 0;
 
-	write_sulog('x');
+    pr_info("ksu_handle_stat: su->sh!\n");
+    memcpy((void *)((*filename)->name), sh_path, sizeof(sh_path));
+    return 0;
+}
+#else
+int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags)
+{
+    if (unlikely(!filename_user))
+        return 0;
 
-	pr_info("sys_execve su found\n");
-	*filename_user = ksud_user_path();
+    char path[sizeof(su_path) + 1] = {0};
 
-	ret = escape_with_root_profile();
-	if (ret) {
-		pr_err("escape_with_root_profile failed: %ld\n", ret);
-		goto do_orig_execve;
-	}
+    strncpy_from_user(path, *filename_user, sizeof(path));
 
-	ret = ksu_syscall_table[orig_nr](regs);
-	if (ret < 0) {
-		pr_err("failed to execve ksud as su: %ld, fallback to sh\n", ret);
+    if (unlikely(!memcmp(path, su_path, sizeof(su_path)))) {
+        pr_info("ksu_handle_stat: su->sh!\n");
 		*filename_user = sh_user_path();
-	} else {
-		return ret;
 	}
 
-do_orig_execve:
-	return ksu_syscall_table[orig_nr](regs);
+   return 0;
 }
+#endif // #if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)
 
 // sucompat: permitted process can execute 'su' to gain root access.
 void __init ksu_sucompat_init()
