diff --git a/kernel/supercall/dispatch.c b/kernel/supercall/dispatch.c
index b5c9bf22..bbb25c6c 100644
--- a/kernel/supercall/dispatch.c
+++ b/kernel/supercall/dispatch.c
@@ -722,11 +722,7 @@ static int do_get_hook_mode(void __user *arg)
 {
     struct ksu_get_hook_mode_cmd cmd = {0};
 
-#ifdef CONFIG_HAVE_SYSCALL_TRACEPOINTS
-    strscpy(cmd.mode, "Tracepoint", sizeof(cmd.mode));
-#else
-    strscpy(cmd.mode, "Kprobes", sizeof(cmd.mode));
-#endif
+    strscpy(cmd.mode, "Inline (SuSFS)", sizeof(cmd.mode));
 
     if (copy_to_user(arg, &cmd, sizeof(cmd))) {
         pr_err("get_hook_mode: copy_to_user failed\n");
