diff --git a/arch/arm64/boot/dts/google/zuma.dtsi b/arch/arm64/boot/dts/google/zuma.dtsi
index 1276e5f..c6ef563 100644
--- a/arch/arm64/boot/dts/google/zuma.dtsi
+++ b/arch/arm64/boot/dts/google/zuma.dtsi
@@ -688,7 +688,8 @@
 	eh: eh@16D00000 {
 		compatible = "google,eh";
 		reg = <0x0 0x16D00000 0x1000>;
-		interrupts = <GIC_SPI IRQ_EH_0_EH IRQ_TYPE_LEVEL_HIGH>;
+		interrupts = <GIC_SPI IRQ_EH_0_EH IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_SPI IRQ_EH_1_EH IRQ_TYPE_LEVEL_HIGH>;
 		clocks = <&clock DOUT_CLK_EH_NOCP>;
 		clock-names = "eh-clock";
 		power-domains = <&pd_eh>;
