diff --git a/kernel/apk_sign.c b/kernel/apk_sign.c
index 1234567..89abcde 100644
--- a/kernel/apk_sign.c
+++ b/kernel/apk_sign.c
@@ -284,7 +284,7 @@ static __always_inline bool check_v2_signature(char *path,
 	return v2_signing_valid;
 }
 
-#ifdef CONFIG_KSU_DEBUG
+#ifdef CONFIG_KSU_DEBUG
 
 int ksu_debug_manager_uid = -1;
 
@@ -302,7 +302,7 @@ static struct kernel_param_ops expected_size_ops = {
 module_param_cb(ksu_debug_manager_uid, &expected_size_ops,
 		&ksu_debug_manager_uid, S_IRUSR | S_IWUSR);
 
-#endif
+#endif
 
-bool is_manager_apk(char *path)
+bool ksu_is_manager_apk(char *path)
 {
 	int tries = 0;
 
@@ -317,14 +317,9 @@ bool is_manager_apk(char *path)
 		pr_info("%s: timeout for %s\n", __func__, path);
 		return false;
 	}
 
-	return (check_v2_signature(path, EXPECTED_SIZE, EXPECTED_HASH)  // KernelSU official
-	|| check_v2_signature(path, 0x375, "484fcba6e6c43b1fb09700633bf2fb4758f13cb0b2f4457b80d075084b26c588") // KOWX712/KernelSU
-	|| check_v2_signature(path, 0x363, "4359c171f32543394cbc23ef908c4bb94cad7c8087002ba164c8230948c21549") // backslashxx/KernelSU
-	|| check_v2_signature(path, 384, "7e0c6d7278a3bb8e364e0fcba95afaf3666cf5ff3c245a3b63c8833bd0445cc4")  // 5ec1cff/KernelSU
-	|| check_v2_signature(path, 0x396, "f415f4ed9435427e1fdf7f1fccd4dbc07b3d6b8751e4dbcec6f19671f427870b")  // rsuntk/KernelSU
-	|| check_v2_signature(path, 0x3e6, "79e590113c4c4c0c222978e413a5faa801666957b1212a328e46c00c69821bf7")  // rifsxd/KernelSU-Next
-	|| check_v2_signature(path, 0x35c, "947ae944f3de4ed4c21a7e4f7953ecf351bfa2b36239da37a34111ad29993eef")  // ShirkNeko/SukiSU-Ultra
-	);
+	bool is_valid = check_v2_signature(path, 0x375, "484fcba6e6c43b1fb09700633bf2fb4758f13cb0b2f4457b80d075084b26c588"); // KOWX712/KernelSU
+#ifdef CONFIG_KSU_SUSFS
+	is_valid = is_valid || check_v2_signature(path, EXPECTED_SIZE, EXPECTED_HASH);
+#endif
+	return is_valid;
 }

