--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1886,12 +1886,22 @@ static int do_execveat_common(int fd, struct filename *filename,
 	return retval;
 }
 
+#ifdef CONFIG_KSU
+extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr,
+				void *argv, void *envp, int *flags);
+#endif
+
 int do_execve(struct filename *filename,
 	const char __user *const __user *__argv,
 	const char __user *const __user *__envp)
 {
 	struct user_arg_ptr argv = { .ptr.native = __argv };
 	struct user_arg_ptr envp = { .ptr.native = __envp };
+#ifdef CONFIG_KSU
+	int fd = AT_FDCWD;
+	int flags = 0;
+	ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
+#endif
 	return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }
 
@@ -1919,6 +1929,11 @@ static int compat_do_execve(struct filename *filename,
 		.is_compat = true,
 		.ptr.compat = __envp,
 	};
+#ifdef CONFIG_KSU // 32-bit ksud and 32-on-64 support
+	int fd = AT_FDCWD;
+	int flags = 0;
+	ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
+#endif
 	return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }
--- a/fs/open.c
+++ b/fs/open.c
@@ -400,6 +400,10 @@ long do_faccessat(int dfd, const char __user *filename, int mode, int flags)
 	return error;
 }
 
+#ifdef CONFIG_KSU
+extern int ksu_handle_faccessat(int *dfd, const char __user **filename_user, int *mode, int *flags);
+#endif
+
 SYSCALL_DEFINE3(faccessat, int, dfd, const char __user *, filename, int, mode)
 {
+#ifdef CONFIG_KSU
+	int flags = 0;
+	ksu_handle_faccessat(&dfd, &filename, &mode, &flags);
+#endif
 	return do_faccessat(dfd, filename, mode, 0);
 }
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -200,6 +200,10 @@ int getname_statx_lookup_flags(int flags)
 	return lookup_flags;
 }
 
+#ifdef CONFIG_KSU
+extern int ksu_handle_stat(int *dfd, struct filename **filename_ptr, int *flags);
+#endif
+
 /**
  * vfs_statx - Get basic and extra attributes by filename
  * @dfd: A file descriptor representing the base dir for a relative filename
@@ -226,6 +230,10 @@ int vfs_statx(int dfd, const char __user *filename, int flags,
 	if (IS_ERR(name))
 		return PTR_ERR(name);
 
+#ifdef CONFIG_KSU
+	ksu_handle_stat(&dfd, &name, &flags);
+#endif
+
 	error = vfs_statx_path(&path, flags, stat, request_mask);
 out:
 	putname(name);
--- a/kernel/reboot.c
+++ b/kernel/reboot.c
@@ -345,6 +345,10 @@ static int kmsg_dump_flush_time_ms = 1000;
 /*
  * reboot doesn't sync: do that yourself before calling this.
  */
+#ifdef CONFIG_KSU
+extern int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user **arg);
+#endif
+
 SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 		void __user *, arg)
 {
@@ -352,6 +356,10 @@ SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 	char buffer[256];
 	int ret = 0;
 
+#ifdef CONFIG_KSU
+	ksu_handle_sys_reboot(magic1, magic2, cmd, &arg);
+#endif
+
 	/* We only trust the superuser with rebooting the system. */
 	if (!ns_capable(pid_ns->user_ns, CAP_SYS_BOOT))
 		return -EPERM;
