diff --git a/include/linux/mount.h b/include/linux/mount.h
index 23baecab7cae..updated 100644
--- a/include/linux/mount.h
+++ b/include/linux/mount.h
@@ -76,7 +76,8 @@ struct vfsmount {
 	ANDROID_KABI_RESERVE(1);
 	ANDROID_KABI_RESERVE(2);
-	ANDROID_KABI_RESERVE(3);
+	ANDROID_KABI_USE(3, void *data);  /* SUSFS: mount private data */
+#ifdef CONFIG_KSU_SUSFS
 	ANDROID_KABI_USE(4, u64 susfs_mnt_id_backup);
 #else
 	ANDROID_KABI_RESERVE(4);
 #endif
} __randomize_layout;
