--- a/kernel/ksud.c
+++ b/kernel/ksud.c
@@ -48,6 +48,8 @@ static struct work_struct stop_input_hook_work;
 #else
 bool ksu_vfs_read_hook __read_mostly = true;
 bool ksu_input_hook __read_mostly = true;
+EXPORT_SYMBOL(ksu_vfs_read_hook);
+EXPORT_SYMBOL(ksu_input_hook);
 #endif
 bool ksu_execveat_hook __read_mostly = true;
 
@@ -616,7 +618,9 @@ void ksu_ksud_init()
 void ksu_ksud_exit()
 {
 #ifdef CONFIG_KPROBES
+#ifdef MODULE
 	unregister_kprobe(&execve_kp);
+#endif
 	// this should be done before unregister vfs_read_kp
 	// unregister_kprobe(&vfs_read_kp);
  
 	unregister_kprobe(&input_event_kp);

