commit 8e023421f3d2274d6d26d0c29cffdf711e700456
Author: debiansid <debiansid@gmail.com>
Date:   Sat Nov 1 19:01:57 2025 +0800

    fix kusd.c

diff --git a/kernel/ksud.c b/kernel/ksud.c
index a998d2a9..d408c14e 100644
--- a/kernel/ksud.c
+++ b/kernel/ksud.c
@@ -52,9 +52,12 @@ static void stop_input_hook();
 static struct work_struct stop_vfs_read_work;
 static struct work_struct stop_execve_hook_work;
 static struct work_struct stop_input_hook_work;
-#else
+#endif
+#if !defined(CONFIG_KSU_KPROBES_HOOK)
 bool ksu_vfs_read_hook __read_mostly = true;
 bool ksu_input_hook __read_mostly = true;
+EXPORT_SYMBOL(ksu_vfs_read_hook);
+EXPORT_SYMBOL(ksu_input_hook);
 #endif
 bool ksu_execveat_hook __read_mostly = true;
 
@@ -283,7 +286,7 @@ static ssize_t read_iter_proxy(struct kiocb *iocb, struct iov_iter *to)
 int ksu_handle_vfs_read(struct file **file_ptr, char __user **buf_ptr,
 			size_t *count_ptr, loff_t **pos)
 {
-#ifndef CONFIG_KPROBES
+#if !defined(CONFIG_KSU_KPROBES_HOOK)
 	if (!ksu_vfs_read_hook) {
 		return 0;
 	}
@@ -396,7 +399,7 @@ static bool is_volumedown_enough(unsigned int count)
 int ksu_handle_input_handle_event(unsigned int *type, unsigned int *code,
 				  int *value)
 {
-#ifndef CONFIG_KPROBES
+#if !defined(CONFIG_KSU_KPROBES_HOOK)
 	if (!ksu_input_hook) {
 		return 0;
 	}
@@ -604,7 +607,9 @@ void ksu_ksud_init()
 void ksu_ksud_exit()
 {
 #ifdef CONFIG_KPROBES
+#ifdef MODULE
 	unregister_kprobe(&execve_kp);
+#endif
 	// this should be done before unregister vfs_read_kp
 	// unregister_kprobe(&vfs_read_kp);
 	unregister_kprobe(&input_event_kp);
