diff --git a/kernel/feature/sucompat.c b/kernel/feature/sucompat.c
index 2ffd10d0..4ad8e056 100644
--- a/kernel/feature/sucompat.c
+++ b/kernel/feature/sucompat.c
@@ -24,27 +24,39 @@
 #include "klog.h" // IWYU pragma: keep
 #include "runtime/ksud.h"
 #include "feature/sucompat.h"
+#include "feature/adb_root.h"
 #include "policy/app_profile.h"
 #include "hook/syscall_hook.h"
 #include "sulog/event.h"
-#include "ksu.h"
-#include "util.h"
+#include "uapi/sulog.h"
 
 #define SU_PATH "/system/bin/su"
 #define SH_PATH "/system/bin/sh"
 
-bool ksu_su_compat_enabled __read_mostly = true;
+static const char sh_path[] = SH_PATH;
+static const char su_path[] = SU_PATH;
+static const char ksud_path[] = KSUD_PATH;
+
+DEFINE_STATIC_KEY_TRUE(ksu_su_compat_enabled);
 
 static int su_compat_feature_get(u64 *value)
 {
-	*value = ksu_su_compat_enabled ? 1 : 0;
+    if (static_key_enabled(&ksu_su_compat_enabled)) {
+        *value = 1;
+	} else {
+        *value = 0;
+	}
 	return 0;
 }
 
 static int su_compat_feature_set(u64 value)
 {
 	bool enable = value != 0;
-	ksu_su_compat_enabled = enable;
+    if (enable) {
+        static_branch_enable(&ksu_su_compat_enabled);
+	} else {
+        static_branch_disable(&ksu_su_compat_enabled);
+	}
 	pr_info("su_compat: set to %d\n", enable);
 	return 0;
 }
@@ -79,180 +91,201 @@ static char __user *ksud_user_path(void)
 	return userspace_stack_buffer(ksud_path, sizeof(ksud_path));
 }
 
-static char __user *empty_user_path(void)
+static const char __user *get_user_arg_ptr(struct user_arg_ptr argv, int nr)
 {
-	return userspace_stack_buffer("", sizeof(""));
-}
+    const char __user *native;
 
-static const char su_path[] = SU_PATH;
+#ifdef CONFIG_COMPAT
+    if (unlikely(argv.is_compat)) {
+        compat_uptr_t compat;
 
-static bool is_ksud_exists()
-{
-	struct path path;
+        if (get_user(compat, argv.ptr.compat + nr))
+            return ERR_PTR(-EFAULT);
 
-	if (kern_path(KSUD_PATH, 0, &path) < 0) {
-		return false;
+        return compat_ptr(compat);
 	}
-	path_put(&path);
-	return true;
+#endif
+
+    if (get_user(native, argv.ptr.native + nr))
+        return ERR_PTR(-EFAULT);
+
+    return native;
 }
 
-long ksu_handle_faccessat_sucompat(int orig_nr, struct pt_regs *regs)
-{
-	const char __user **filename_user, *orig_filename;
-	long ret;
-	const struct cred *old_cred;
+/*
+ * return 0 -> No further checks should be required afterwards
+ * return non-zero -> Further checks should be continued afterwards
+ */
+int ksu_handle_execveat_init(struct filename *filename, struct user_arg_ptr *argv_user, struct user_arg_ptr *envp_user) {
+    int ret;
 
-	if (!ksu_is_allow_uid_for_current(current_uid().val)) {
-		goto do_orig_facessat;
-	}
+    if (current->pid == 1)
+        return -EINVAL;
+
+    if (!is_init(get_current_cred()))
+        return -EINVAL;
 
-	filename_user = (const char __user **)&PT_REGS_PARM2(regs);
-
-	char path[sizeof(su_path) + 1];
-	memset(path, 0, sizeof(path));
-	strncpy_from_user_nofault(path, *filename_user, sizeof(path));
-
-	if (unlikely(!memcmp(path, su_path, sizeof(su_path)))) {
-		old_cred = override_creds(ksu_cred);
-		if (is_ksud_exists()) {
-			ksu_compat_sulog('a');
-			pr_info("faccessat su->ksud!\n");
-			orig_filename = *filename_user;
-			*filename_user = ksud_user_path();
-			ret = ksu_syscall_table[orig_nr](regs);
-			revert_creds(old_cred);
-			*filename_user = orig_filename;
+    if (unlikely(!strcmp(filename->name, KSUD_PATH))) {
+        const char __user *argv_user_ptr = get_user_arg_ptr(*argv_user, 0);
+        struct ksu_sulog_pending_event *pending_sucompat = NULL;
+
+        pr_info("hook_manager: escape to root for init executing ksud: %d\n", current->pid);
+        pending_sucompat = ksu_sulog_capture_sucompat(filename->name, argv_user, GFP_KERNEL);
+        ret = escape_to_root_for_init();
+        if (ret) {
+            pr_err("escape_to_root_for_init() failed: %d\n", ret);
 			return ret;
-		} else {
-			revert_creds(old_cred);
 		}
+        if (!argv_user_ptr || IS_ERR(argv_user_ptr)) {
+            pr_err("!argv_user_ptr || IS_ERR(argv_user_ptr)\n");
+            return 0;
+        }
+        ksu_sulog_emit_pending(pending_sucompat, ret, GFP_KERNEL);
+        return 0;
 	}
 
-do_orig_facessat:
-	return ksu_syscall_table[orig_nr](regs);
+    if (likely(!strstr(filename->name, "/app_process") && !strstr(filename->name, "/adbd"))) {
+        pr_info("susfs: mark no sucompat checks for pid: '%d', exec: '%s'\n", current->pid, filename->name);
+        susfs_set_current_proc_umounted();
+        return 0;
+    }
+
+#ifdef CONFIG_COMPAT
+    if (unlikely(envp_user->is_compat))
+        ret = ksu_adb_root_handle_execve(filename->name, (void ***)&envp_user->ptr.compat);
+    else
+        ret = ksu_adb_root_handle_execve(filename->name, (void ***)&envp_user->ptr.native);
+#else
+        ret = ksu_adb_root_handle_execve(filename->name, (void ***)&envp_user->ptr.native);
+#endif
+
+    if (ret)
+        pr_err("adb root failed: %d\n", ret);
+
+    return ret;
 }
 
-long ksu_handle_stat_sucompat(int orig_nr, struct pt_regs *regs)
+// the call from execve_handler_pre won't provided correct value for __never_use_argument, use them after fix execve_handler_pre, keeping them for consistence for manually patched code
+int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
+                 void *argv_user, void *envp_user,
+                 int *__never_use_flags)
 {
-	const char __user **filename_user, *orig_filename;
-	long ret;
-	const struct cred *old_cred;
+    struct filename *filename;
+    struct ksu_sulog_pending_event *pending_sucompat = NULL;
+    int ret;
+
+    if (unlikely(!filename_ptr))
+        return 0;
+
+    filename = *filename_ptr;
+    if (IS_ERR(filename))
+        return 0;
+
+    if (!ksu_handle_execveat_init(filename, (struct user_arg_ptr*)argv_user, (struct user_arg_ptr*)envp_user))
+        return 0;
+
+    if (!(__ksu_is_allow_uid_for_current(current_uid().val)))
+        return 0;
+
+    if (likely(memcmp(filename->name, su_path, sizeof(su_path))))
+        return 0;
 
-	if (!ksu_is_allow_uid_for_current(current_uid().val)) {
-		goto do_orig_stat;
+    if (current_chrooted())
+    {
+        pr_err("ksu_handle_execveat_sucompat: su found but NOT allowed! Because current process is running in chrooted environment\n");
+        return 0;
 	}
 
-	filename_user = (const char __user **)&PT_REGS_PARM2(regs);
-
-	char path[sizeof(su_path) + 1];
-	memset(path, 0, sizeof(path));
-	strncpy_from_user_nofault(path, *filename_user, sizeof(path));
-
-	if (unlikely(!memcmp(path, su_path, sizeof(su_path)))) {
-		old_cred = override_creds(ksu_cred);
-		if (is_ksud_exists()) {
-			ksu_compat_sulog('s');
-			pr_info("newfstatat su->ksud!\n");
-			orig_filename = *filename_user;
-			*filename_user = ksud_user_path();
-			ret = ksu_syscall_table[orig_nr](regs);
-			revert_creds(old_cred);
-			*filename_user = orig_filename;
-			return ret;
-		} else {
-			revert_creds(old_cred);
-		}
+    pr_info("ksu_handle_execveat_sucompat: su found\n");
+
+    memcpy((void *)filename->name, ksud_path, sizeof(ksud_path));
+
+    pending_sucompat = ksu_sulog_capture_sucompat(filename->name, (struct user_arg_ptr*)argv_user, GFP_KERNEL);
+
+    ret = escape_with_root_profile();
+    if (ret)
+        pr_err("escape_with_root_profile() failed: %d\n", ret);
+
+    const char __user *argv_user_ptr = get_user_arg_ptr(*((struct user_arg_ptr*)argv_user), 0);
+    if (!argv_user_ptr || IS_ERR(argv_user_ptr)) {
+        pr_err("!argv_user_ptr || IS_ERR(argv_user_ptr)\n");
+        return 0;
 	}
 
-do_orig_stat:
-	return ksu_syscall_table[orig_nr](regs);
+    ksu_sulog_emit_pending(pending_sucompat, ret, GFP_KERNEL);
+    return 0;
+}
+
+extern struct static_key_true is_first_zygote;
+
+int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv,
+            void *envp, int *flags)
+{
+    if (static_branch_unlikely(&is_first_zygote))
+        (void)ksu_handle_execveat_ksud(fd, filename_ptr, argv, envp, flags);
+
+    return ksu_handle_execveat_sucompat(fd, filename_ptr, argv, envp,
+                        flags);
 }
 
-long ksu_handle_execve_sucompat(const char __user **filename_user, int orig_nr, struct pt_regs *regs)
+int ksu_handle_faccessat(int *dfd, const char __user **filename_user, int *mode,
+             int *__unused_flags)
 {
-	const char __user *fn;
-	const char __user *const __user *argv_user = (const char __user *const __user *)PT_REGS_PARM2(regs);
-	struct ksu_sulog_pending_event *pending_sucompat = NULL;
-	char path[sizeof(su_path) + 1];
-	long ret, orig_regs[5];
-	unsigned long addr;
-	int tmp_fd;
-	struct file *ksud_file;
-	const struct cred *old_cred;
-
-	if (unlikely(!filename_user))
-		goto do_orig_execve;
-
-	if (!ksu_is_allow_uid_for_current(current_uid().val))
-		goto do_orig_execve;
-
-	addr = untagged_addr((unsigned long)*filename_user);
-	fn = (const char __user *)addr;
-	memset(path, 0, sizeof(path));
-	ret = strncpy_from_user(path, fn, sizeof(path));
-
-	if (ret < 0) {
-		pr_warn("Access filename when execve failed: %ld", ret);
-		goto do_orig_execve;
+    char path[sizeof(su_path) + 1] = {0};
+	strncpy_from_user(path, *filename_user, sizeof(path));
+
+    if (unlikely(!memcmp(path, su_path, sizeof(su_path)))) {
+        if (current_chrooted())
+        {
+            pr_err("ksu_handle_faccessat: su found but NOT allowed! Because current process is running in chrooted environment\n");
+            return 0;
+        }
+        pr_info("ksu_handle_faccessat: su->sh!\n");
+        *filename_user = sh_user_path();
 	}
 
-	if (likely(memcmp(path, su_path, sizeof(su_path))))
-		goto do_orig_execve;
+    return 0;
+}
 
-	ksu_compat_sulog('x');
-	pr_info("sys_execve su found\n");
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)
+int ksu_handle_stat(int *dfd, struct filename **filename, int *flags) {
+    if (unlikely(IS_ERR(*filename) || (*filename)->name == NULL))
+        return 0;
 
-	tmp_fd = get_unused_fd_flags(O_CLOEXEC);
-	if (tmp_fd < 0) {
-		pr_err("alloc tmp fd err: %d\n", tmp_fd);
-		goto do_orig_execve;
-	}
+    if (likely(memcmp((*filename)->name, su_path, sizeof(su_path))))
+        return 0;
 
-	old_cred = override_creds(ksu_cred);
-	ksud_file = filp_open(KSUD_PATH, O_PATH, 0);
-	revert_creds(old_cred);
-	if (IS_ERR(ksud_file)) {
-		pr_err("open ksud err: %ld\n", PTR_ERR(ksud_file));
-		put_unused_fd(tmp_fd);
-		goto do_orig_execve;
+    if (current_chrooted())
+    {
+        pr_err("ksu_handle_stat: su found but NOT allowed! Because current process is running in chrooted environment\n");
+        return 0;
 	}
+    pr_info("ksu_handle_stat: su->sh!\n");
+    memcpy((void *)((*filename)->name), sh_path, sizeof(sh_path));
+    return 0;
+}
+#else
+int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags)
+{
+    if (unlikely(!filename_user))
+        return 0;
 
-	fd_install(tmp_fd, ksud_file);
-
-	pending_sucompat = ksu_sulog_capture_sucompat(*filename_user, argv_user, GFP_KERNEL);
-	// execve(file, argv, environ)
-	// execveat(fd, file, argv, environ, flags)
-	orig_regs[0] = regs->__PT_PARM1_REG;
-	orig_regs[1] = regs->__PT_PARM2_REG;
-	orig_regs[2] = regs->__PT_PARM3_REG;
-	orig_regs[3] = regs->__PT_SYSCALL_PARM4_REG;
-	orig_regs[4] = regs->__PT_PARM5_REG;
-	regs->__PT_PARM5_REG = AT_EMPTY_PATH;
-	regs->__PT_SYSCALL_PARM4_REG = regs->__PT_PARM3_REG;
-	regs->__PT_PARM3_REG = regs->__PT_PARM2_REG;
-	regs->__PT_PARM2_REG = empty_user_path();
-	regs->__PT_PARM1_REG = tmp_fd;
-
-	ret = escape_with_root_profile();
-	if (ret) {
-		pr_err("escape_with_root_profile failed: %ld\n", ret);
-	}
-	ksu_sulog_emit_pending(pending_sucompat, ret, GFP_KERNEL);
-
-	ret = ksu_syscall_table[__NR_execveat](regs);
-	if (ret < 0) {
-		ksu_close_fd(tmp_fd);
-		regs->__PT_PARM1_REG = orig_regs[0];
-		regs->__PT_PARM2_REG = orig_regs[1];
-		regs->__PT_PARM3_REG = orig_regs[2];
-		regs->__PT_SYSCALL_PARM4_REG = orig_regs[3];
-		regs->__PT_PARM5_REG = orig_regs[4];
+    char path[sizeof(su_path) + 1] = {0};
+
+	strncpy_from_user(path, *filename_user, sizeof(path));
+
+    if (unlikely(!memcmp(path, su_path, sizeof(su_path)))) {
+        if (current_chrooted())
+        {
+            pr_err("ksu_handle_stat: su found but NOT allowed! Because current process is running in chrooted environment\n");
+            return 0;
+        }
+        pr_info("ksu_handle_stat: su->sh!\n");
+        *filename_user = sh_user_path();
 	}
-	return ret;
 
-do_orig_execve:
-	return ksu_syscall_table[orig_nr](regs);
+    return 0;
 }
+#endif // #if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)
 
 // sucompat: permitted process can execute 'su' to gain root access.
