diff --git a/kernel/app_profile.c b/kernel/app_profile.c
index 490de8d..f859057 100644
--- a/kernel/app_profile.c
+++ b/kernel/app_profile.c
@@ -11,7 +11,9 @@
 #include "app_profile.h"
 #include "klog.h" // IWYU pragma: keep
 #include "selinux/selinux.h"
+#ifndef CONFIG_KSU_SUSFS
 #include "syscall_hook_manager.h"
+#endif // #ifndef CONFIG_KSU_SUSFS
 
 static struct group_info root_groups = { .usage = ATOMIC_INIT(2) };
 
@@ -130,10 +132,11 @@ void escape_with_root_profile(void)
     spin_unlock_irq(&current->sighand->siglock);
 
     setup_selinux(profile->selinux_domain);
-
+#ifndef CONFIG_KSU_SUSFS
     for_each_thread (p, t) {
         ksu_set_task_tracepoint_flag(t);
     }
+#endif // #ifndef CONFIG_KSU_SUSFS
 }
 
 void escape_to_root_for_init(void)
