diff --git a/kernel/ksu.c b/kernel/ksu.c
index 06074beb..c8939344 100644
--- a/kernel/ksu.c
+++ b/kernel/ksu.c
@@ -32,14 +32,19 @@ int __init kernelsu_init(void)
 	ksu_feature_init();
 
 	ksu_supercalls_init();
-
+#ifndef CONFIG_KSU_SUSFS
 	ksu_syscall_hook_manager_init();
-
+#endif // #ifndef CONFIG_KSU_SUSFS
 	ksu_allowlist_init();
 
 	ksu_throne_tracker_init();
+#ifdef CONFIG_KSU_SUSFS
+    susfs_init();
+#endif // #ifdef CONFIG_KSU_SUSFS
 
+#ifndef CONFIG_KSU_SUSFS
 	ksu_ksud_init();
+#endif // #ifndef CONFIG_KSU_SUSFS
 
 #ifdef MODULE
 #ifndef CONFIG_KSU_DEBUG
@@ -57,11 +62,11 @@ void kernelsu_exit(void)
 	ksu_throne_tracker_exit();
 
 	ksu_observer_exit();
-
+#ifndef CONFIG_KSU_SUSFS
 	ksu_ksud_exit();
 
 	ksu_syscall_hook_manager_exit();
-
+#endif // #ifndef CONFIG_KSU_SUSFS
 	ksu_supercalls_exit();
 
 	ksu_feature_exit();
