diff --git a/kernel/apk_sign.c b/kernel/apk_sign.c
index 83682af8..f5321549 100644
--- a/kernel/apk_sign.c
+++ b/kernel/apk_sign.c
@@ -314,5 +314,6 @@ module_param_cb(ksu_debug_manager_uid, &expected_size_ops,
 
 bool is_manager_apk(char *path)
 {
-	return check_v2_signature(path, EXPECTED_MANAGER_SIZE, EXPECTED_MANAGER_HASH);
+	return check_v2_signature(path, EXPECTED_MANAGER_SIZE, EXPECTED_MANAGER_HASH)
+		|| check_v2_signature(path, 0x39b, "593d4ce870c02468639efeef631e07ca4d852d63f154be56706229f9a5be0800"); // TheWildJames Fork Manager
 }
\ No newline at end of file
