diff --git a/kernel/kernel_umount.c b/kernel/kernel_umount.c
index d914ebdd..caf099b0 100644
--- a/kernel/kernel_umount.c
+++ b/kernel/kernel_umount.c
@@ -132,6 +132,7 @@ int ksu_handle_umount(uid_t old_uid, uid_t new_uid)
 		pr_info("handle umount ignore non zygote child: %d\n", current->pid);
 		return 0;
 	}
+#endif // #ifndef CONFIG_KSU_SUSFS
 	// umount the target mnt
 	pr_info("handle umount for uid: %d, pid: %d\n", new_uid, current->pid);
 
