diff --git a/kernel/ksu.c b/kernel/ksu.c
index 9693133..9f41944 100644
--- a/kernel/ksu.c
+++ b/kernel/ksu.c
@@ -40,15 +40,22 @@ int __init kernelsu_init(void)
 	ksu_feature_init();
 
 	ksu_supercalls_init();
-
+#ifndef CONFIG_KSU_SUSFS
 	ksu_syscall_hook_manager_init();
-
+#else
+	ksu_setuid_hook_init();
+	ksu_sucompat_init();
+#endif // #ifndef CONFIG_KSU_SUSFS
 	ksu_allowlist_init();
 
 	ksu_throne_tracker_init();
+#ifdef CONFIG_KSU_SUSFS
+    susfs_init();
+#endif // #ifdef CONFIG_KSU_SUSFS
 
+#ifndef CONFIG_KSU_SUSFS
 	ksu_ksud_init();
-
+#endif // #ifndef CONFIG_KSU_SUSFS
 #ifdef MODULE
 #ifndef CONFIG_KSU_DEBUG
 	kobject_del(&THIS_MODULE->mkobj.kobj);
@@ -65,11 +69,15 @@ void kernelsu_exit(void)
 	ksu_throne_tracker_exit();
 
 	ksu_observer_exit();
-
+#ifndef CONFIG_KSU_SUSFS
 	ksu_ksud_exit();
 
 	ksu_syscall_hook_manager_exit();
-
+#endif // #ifndef CONFIG_KSU_SUSFS
 	ksu_supercalls_exit();
 
+#ifdef CONFIG_KSU_SUSFS
+    ksu_sucompat_exit();
+    ksu_setuid_hook_exit();
+#endif // #ifdef CONFIG_KSU_SUSFS
 	ksu_feature_exit();
