diff --git a/kernel/ksu.c b/kernel/ksu.c
index 02864fd..cc72b22 100644
--- a/kernel/ksu.c
+++ b/kernel/ksu.c
@@ -43,15 +43,21 @@ int __init kernelsu_init(void)
 	ksu_feature_init();
 
 	ksu_supercalls_init();
-
+#ifndef CONFIG_KSU_SUSFS
 	ksu_syscall_hook_manager_init();
-
+#else
+	ksu_setuid_hook_init();
+	ksu_sucompat_init();
+#endif // #ifndef CONFIG_KSU_SUSFS
 	ksu_allowlist_init();
 
 	ksu_throne_tracker_init();
 
+#ifdef CONFIG_KSU_SUSFS
+	susfs_init();
+#else
 	ksu_ksud_init();
-
+#endif // #ifdef CONFIG_KSU_SUSFS
 #ifdef MODULE
 #ifndef CONFIG_KSU_DEBUG
 	kobject_del(&THIS_MODULE->mkobj.kobj);
@@ -68,11 +75,11 @@ void kernelsu_exit(void)
 	ksu_throne_tracker_exit();
 
 	ksu_observer_exit();
-
+#ifndef CONFIG_KSU_SUSFS
 	ksu_ksud_exit();
 
 	ksu_syscall_hook_manager_exit();
-
+#endif // #ifndef CONFIG_KSU_SUSFS
 	ksu_supercalls_exit();
 
 	ksu_feature_exit();
