--- a/kernel/ksu.c
+++ b/kernel/ksu.c
@@ -39,15 +39,23 @@
 
 	ksu_feature_init();
 
-	ksu_supercalls_init();
+    ksu_supercalls_init();
 
-	ksu_syscall_hook_manager_init();
+#ifndef CONFIG_KSU_SUSFS
+    ksu_syscall_hook_manager_init();
+#endif // #ifndef CONFIG_KSU_SUSFS
 
-	ksu_allowlist_init();
+    ksu_allowlist_init();
 
-	ksu_throne_tracker_init();
+    ksu_throne_tracker_init();
 
-	ksu_ksud_init();
+#ifdef CONFIG_KSU_SUSFS
+    susfs_init();
+#endif // #ifdef CONFIG_KSU_SUSFS
+
+#ifndef CONFIG_KSU_SUSFS
+    ksu_ksud_init();
+#endif // #ifndef CONFIG_KSU_SUSFS
 
 #ifdef MODULE
 #ifndef CONFIG_KSU_DEBUG
@@ -64,15 +72,17 @@
 
 	ksu_throne_tracker_exit();
 
-	ksu_observer_exit();
+    ksu_observer_exit();
 
-	ksu_ksud_exit();
+#ifndef CONFIG_KSU_SUSFS
+    ksu_ksud_exit();
 
-	ksu_syscall_hook_manager_exit();
+    ksu_syscall_hook_manager_exit();
+#endif // #ifndef CONFIG_KSU_SUSFS
 
-	ksu_supercalls_exit();
+    ksu_supercalls_exit();
 
-	ksu_feature_exit();
+    ksu_feature_exit();
 
 	if (ksu_cred) {
 		put_cred(ksu_cred);
