diff --git a/kernel/supercalls.c b/kernel/supercalls.c
index da39800..1229b02 100644
--- a/kernel/supercalls.c
+++ b/kernel/supercalls.c
@@ -466,7 +466,7 @@ static int do_get_hook_mode(void __user *arg)
 {
 	struct ksu_get_hook_mode_cmd cmd = {0};
 
-	strscpy(cmd.mode, "Kprobes", sizeof(cmd.mode));
+	strscpy(cmd.mode, "Inline (SuSFS)", sizeof(cmd.mode));
 
 	if (copy_to_user(arg, &cmd, sizeof(cmd))) {
 		pr_err("get_hook_mode: copy_to_user failed\n");
