diff --git a/security/selinux/avc.c b/security/selinux/avc.c
index 15e526609b62..xxxxxxxxxxxxx 100644
--- a/security/selinux/avc.c
+++ b/security/selinux/avc.c
@@ -49,7 +49,6 @@
 
 #ifdef CONFIG_KSU_SUSFS
 extern u32 susfs_ksu_sid;
-extern u32 susfs_priv_app_sid;
 bool susfs_is_avc_log_spoofing_enabled = false;
 #endif
 
@@ -740,7 +739,7 @@ static void avc_audit_post_callback(struct audit_buffer *ab, void *a)
 				     &tcontext_len);
 #ifdef CONFIG_KSU_SUSFS
 	if (unlikely(sad->tsid == susfs_ksu_sid && susfs_is_avc_log_spoofing_enabled)) {
 		if (rc)
-			audit_log_format(ab, " tsid=%d", susfs_priv_app_sid);
+			audit_log_format(ab, " tsid=%d", sad->tsid);
 		else
 			audit_log_format(ab, " tcontext=%s", "u:r:priv_app:s0:c512,c768");
 		goto bypass_orig_flow;
