--- ./kernel/apk_sign.c	2025-07-19 15:34:49.962023700 -0400
+++ ./kernel/apk_sign.c	2025-07-19 15:41:30.015681476 -0400
@@ -363,7 +363,7 @@
 
 #endif
 
-bool is_manager_apk(char *path)
+bool ksu_is_manager_apk(char *path)
 {
 	int tries = 0;
 
@@ -388,5 +388,10 @@
 	pr_info("%s: expected size: %u, expected hash: %s\n",
 		path, expected_manager_size, expected_manager_hash);
 
+#ifdef CONFIG_KSU_SUSFS
+	return (check_v2_signature(path, expected_manager_size, expected_manager_hash)
+			|| check_v2_signature(path, 0x39b, "593d4ce870c02468639efeef631e07ca4d852d63f154be56706229f9a5be0800")); // TheWildJames Fork Manager
+#else
 	return check_v2_signature(path, expected_manager_size, expected_manager_hash);
+#endif
 }
