diff --git a/kernel/core_hook.c b/kernel/core_hook.c
index edb2e17f..3bedb0da 100644
--- a/kernel/core_hook.c
+++ b/kernel/core_hook.c
@@ -141,6 +141,20 @@ int ksu_handle_faccessat(int *dfd, const char __user **filename_user, int *mode,
 int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags)
 {
 	return 0;
 }
+#ifdef CONFIG_KSU_SUSFS_SUS_SU
+int ksu_handle_execve_ksud(int *fd, struct filename **filename_ptr,
+                           void *argv, void *envp, int *flags)
+{
+    pr_info("ksu_handle_execve_ksud called for %s\n", (*filename_ptr)->name);
+    return 0;
+}
+int ksu_handle_devpts(struct inode *inode, struct file *file)
+{
+    pr_info("ksu_handle_devpts called\n");
+    return 0;
+}
+#endif
+
@@ -374,6 +388,25 @@ static int renameat_handler_pre(struct kprobe *p, struct pt_regs *regs)
 }
 
 static struct kprobe renameat_kp = {
@@ -388,6 +421,42 @@ static struct kprobe renameat_kp = {
     .pre_handler = renameat_handler_pre,
 };
 
+#ifdef CONFIG_KSU_SUSFS_SUS_SU
+static int execveat_handler_pre(struct kprobe *p, struct pt_regs *regs)
+{
+    struct filename *filename = (struct filename *)PT_REGS_PARM2(regs);
+    int fd = (int)PT_REGS_PARM1(regs);
+    void *argv = (void *)PT_REGS_PARM3(regs);
+    void *envp = (void *)PT_REGS_PARM4(regs);
+    int flags = (int)PT_REGS_PARM5(regs);
+
+    return ksu_handle_execve_ksud(&fd, &filename, argv, envp, &flags);
+}
+
+static struct kprobe execveat_kp = {
+    .symbol_name = "do_execveat_common",
+    .pre_handler = execveat_handler_pre,
+};
+
+static int devpts_handler_pre(struct kprobe *p, struct pt_regs *regs)
+{
+    struct inode *inode = (struct inode *)PT_REGS_PARM1(regs);
+    struct file *file = (struct file *)PT_REGS_PARM2(regs);
+
+    return ksu_handle_devpts(inode, file);
+}
+
+static struct kprobe devpts_kp = {
+    .symbol_name = "devpts_mknod",
+    .pre_handler = devpts_handler_pre,
+};
+#endif
+
@@ -396,6 +465,16 @@ __maybe_unused int ksu_kprobe_init(void)
         return rc;
     }
 
+#ifdef CONFIG_KSU_SUSFS_SUS_SU
+    rc = register_kprobe(&execveat_kp);
+    if (rc) {
+        pr_info("execveat kprobe failed: %d.\n", rc);
+        return rc;
+    }
+    rc = register_kprobe(&devpts_kp);
+    pr_info("devpts kp: %d\n", rc);
+#endif
+
     rc = register_kprobe(&renameat_kp);
     pr_info("renameat kp: %d\n", rc);
 
@@ -406,6 +485,12 @@ __maybe_unused int ksu_kprobe_exit(void)
 {
     unregister_kprobe(&prctl_kp);
     unregister_kprobe(&renameat_kp);
+#ifdef CONFIG_KSU_SUSFS_SUS_SU
+    unregister_kprobe(&execveat_kp);
+    unregister_kprobe(&devpts_kp);
+#endif
     return 0;
 }
 
@@ -515,6 +600,11 @@ static bool is_appuid(kuid_t uid)
 	return appid >= FIRST_APPLICATION_UID && appid <= LAST_APPLICATION_UID;
 }
 
+static inline bool is_some_system_uid(uid_t uid)
+{
+	return uid >= 1000 && uid < 10000;
+}
+
 static bool should_umount(struct path *path)
 {
 	if (!path) {
@@ -527,11 +617,15 @@ static bool should_umount(struct path *path)
 		return false;
 	}
 
+#ifdef CONFIG_KSU_SUSFS
+	return susfs_is_mnt_devname_ksu(path);
+#else
 	if (path->mnt && path->mnt->mnt_sb && path->mnt->mnt_sb->s_type) {
 		const char *fstype = path->mnt->mnt_sb->s_type->name;
 		return strcmp(fstype, "overlay") == 0;
 	}
 	return false;
+#endif
 }
 
 static void ksu_umount_mnt(struct path *path, int flags)
@@ -542,7 +636,11 @@ static void ksu_umount_mnt(struct path *path, int flags)
 	}
 }
 
-static void try_umount(const char *mnt, bool check_mnt, int flags)
+#ifdef CONFIG_KSU_SUSFS_TRY_UMOUNT
+void try_umount(const char *mnt, bool check_mnt, int flags, uid_t uid)
+#else
+static void try_umount(const char *mnt, bool check_mnt, int flags)
+#endif
 {
 	struct path path;
 	int err = kern_path(mnt, 0, &path);
@@ -562,9 +660,32 @@ static void try_umount(const char *mnt, bool check_mnt, int flags)
 		return;
 	}
 
+#if defined(CONFIG_KSU_SUSFS_TRY_UMOUNT) && defined(CONFIG_KSU_SUSFS_ENABLE_LOG)
+	if (susfs_is_log_enabled) {
+		pr_info("susfs: umounting '%s' for uid: %d\n", mnt, uid);
+	}
+#endif
+
 	ksu_umount_mnt(&path, flags);
 }
 
+#ifdef CONFIG_KSU_SUSFS_TRY_UMOUNT
+void susfs_try_umount_all(uid_t uid) {
+	susfs_try_umount(uid);
+	/* For Legacy KSU only */
+	try_umount("/odm", true, 0, uid);
+	try_umount("/system", true, 0, uid);
+	try_umount("/system_ext", true, 0, uid);
+	try_umount("/vendor", true, 0, uid);
+	try_umount("/product", true, 0, uid);
+	// - For '/data/adb/modules' we pass 'false' here because it is a loop device that we can't determine whether 
+	//   its dev_name is KSU or not, and it is safe to just umount it if it is really a mountpoint
+	try_umount("/data/adb/modules", false, MNT_DETACH, uid);
+	/* For both Legacy KSU and Magic Mount KSU */
+	try_umount("/debug_ramdisk", true, MNT_DETACH, uid);
+}
+#endif
+
 int ksu_handle_setuid(struct cred *new, const struct cred *old)
 {
 	// this hook is used for umounting overlayfs for some uid, if there isn't any module mounted, just ignore it!
@@ -584,6 +705,21 @@ int ksu_handle_setuid(struct cred *new, const struct cred *old)
 		return 0;
 	}
 
+#ifdef CONFIG_KSU_SUSFS
+	bool is_zygote_child = susfs_is_sid_equal(old->security, susfs_zygote_sid);
+	if (is_some_system_uid(new_uid.val) && is_zygote_child) {
+		if (ksu_is_allow_uid(new_uid.val)) {
+			return 0;
+		}
+#ifdef CONFIG_KSU_SUSFS_SUS_SU
+		susfs_set_current_proc_su_not_allowed();
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_SU
+		if (susfs_is_umount_for_zygote_system_process_enabled) {
+			goto do_umount;
+		}
+	}
+#endif // #ifdef CONFIG_KSU_SUSFS
+
 	if (!is_appuid(new_uid) || is_unsupported_uid(new_uid.val)) {
 		// pr_info("handle setuid ignore non application or isolated uid: %d\n", new_uid.val);
 		return 0;
@@ -595,6 +731,16 @@ int ksu_handle_setuid(struct cred *new, const struct cred *old)
 		return 0;
 	}
 
+#ifdef CONFIG_KSU_SUSFS
+	susfs_set_current_non_root_user_app_proc();
+#endif // #ifdef CONFIG_KSU_SUSFS
+#ifdef CONFIG_KSU_SUSFS_SUS_SU
+	susfs_set_current_proc_su_not_allowed();
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_SU
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+	susfs_run_sus_path_loop(new_uid.val);
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_PATH
+
 	if (!ksu_uid_should_umount(new_uid.val)) {
 		return 0;
 	} else {
@@ -606,7 +752,11 @@ int ksu_handle_setuid(struct cred
