--- a/fs/open.c 2026-04-06 21:41:37.218666167 +0800
+++ b/fs/open.c	2026-04-06 22:24:41.206965183 +0800
@@ -1421,8 +1421,29 @@
 		return PTR_ERR(tmp);
 
 	fd = get_unused_fd_flags(how->flags);
+#ifdef CONFIG_KSU_SUSFS_OPEN_REDIRECT
+    retry:
+#endif
+
 	if (fd >= 0) {
 		struct file *f = do_filp_open(dfd, tmp, &op);
+#ifdef CONFIG_KSU_SUSFS_OPEN_REDIRECT
+		/* Check if this inode should be redirected to a different path */
+		if (!is_inode_open_redirect && f && !IS_ERR(f)) {
+			struct inode *inode = file_inode(f);
+			if (SUSFS_IS_INODE_OPEN_REDIRECT_WITHOUT_UID_CHECK(inode)) {
+				fake_filename = susfs_open_redirect_spoof_do_sys_openat(inode);
+				if (fake_filename && !IS_ERR(fake_filename)) {
+					is_inode_open_redirect = true;
+					filp_close(f, NULL); // Close the 'wrong' file
+					putname(tmp);        // Free the old path name
+					tmp = fake_filename; // Swap to the spoofed path
+					goto retry;          // Try opening again with the new path
+				}
+			}
+		}
+#endif
+
 		if (IS_ERR(f) && !libperfmgr_redirect(&f, dfd, tmp, &op, how)) {
 			put_unused_fd(fd);
 			fd = PTR_ERR(f);
