--- a/security/selinux/avc.c
+++ b/security/selinux/avc.c
@@ -47,6 +47,10 @@
 #undef CREATE_TRACE_POINTS
 #include <trace/hooks/avc.h>
 
+#if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+extern void ksu_slow_avc_audit(u32 *tsid);
+#endif
+
 struct avc_entry {
        u32                     ssid;
        u32                     tsid;
@@ -781,6 +785,10 @@
 {
        struct common_audit_data stack_data;
        struct selinux_audit_data sad;
+       
+       #if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+       ksu_slow_avc_audit(&tsid);
+#endif
 
        if (WARN_ON(!tclass || tclass >= ARRAY_SIZE(secclass_map)))
                return -EINVAL;

--- a/security/selinux/selinuxfs.c
+++ b/security/selinux/selinuxfs.c
@@ -43,6 +43,10 @@
 #include "conditional.h"
 #include "ima.h"
 
+#if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+extern void ksu_sel_write_context(struct file **file, char **buf, size_t *size);
+#endif
+
 enum sel_inos {
        SEL_ROOT_INO = 2,
        SEL_LOAD,       /* load policy */
@@ -888,7 +892,9 @@
        data = simple_transaction_get(file, buf, size);
        if (IS_ERR(data))
                return PTR_ERR(data);
-
+       #if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+       ksu_sel_write_context(&file, &data, &size);
+       #endif
        rv = write_op[ino](file, data, size);
        if (rv > 0) {
                simple_transaction_set(file, rv);
