--- a/security/selinux/avc.c
+++ b/security/selinux/avc.c
@@ -769,6 +769,10 @@
 	}
 }
 
+#if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+extern void ksu_slow_avc_audit(u32 *tsid);
+#endif
+
 /*
  * This is the slow part of avc audit with big stack footprint.
  * Note that it is non-blocking and can be called from under
@@ -784,7 +788,10 @@
 
 	if (WARN_ON(!tclass || tclass >= ARRAY_SIZE(secclass_map)))
 		return -EINVAL;
-
+	
+#if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+	ksu_slow_avc_audit(&tsid);
+#endif
 	if (!a) {
 		a = &stack_data;
 		a->type = LSM_AUDIT_DATA_NONE;
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c	
@@ -6443,6 +6443,10 @@
 	return error;
 }
 
+#ifdef CONFIG_KSU
+extern int ksu_hide_setprocattr(const char *name, void *value, size_t size);
+#endif
+
 static int selinux_setprocattr(const char *name, void *value, size_t size)
 {
 	struct task_security_struct *tsec;
@@ -6451,6 +6455,10 @@
 	int error;
 	char *str = value;
 
+#ifdef CONFIG_KSU
+       ksu_hide_setprocattr(name, value, size);
+#endif
+       
 	/*
 	 * Basic control over ability to set these attributes at all.
 	 */
--- a/security/selinux/selinuxfs.c
+++ b/security/selinux/selinuxfs.c
@@ -876,6 +876,10 @@
 	[SEL_CONTEXT] = sel_write_context,
 };
 
+#if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+extern void ksu_sel_write_context(struct file **file, char **buf, size_t *size);
+#endif
+
 static ssize_t selinux_transaction_write(struct file *file, const char __user *buf, size_t size, loff_t *pos)
 {
 	ino_t ino = file_inode(file)->i_ino;
@@ -889,6 +893,10 @@
 	if (IS_ERR(data))
 		return PTR_ERR(data);
 
+#if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+	ksu_sel_write_context(&file, &data, &size);
+#endif
+
 	rv = write_op[ino](file, data, size);
 	if (rv > 0) {
 		simple_transaction_set(file, rv);
